x-pack/plugins/ml/public/application/contexts/storage/storage_context.tsx

```diff
@@ -106,21 +106,32 @@ export const MlStorageContextProvider: FC = ({ children }) => {
  * @param key
  * @param initValue
  */
-export function useStorage<K extends MlStorageKey>(
+export function useStorage<K extends MlStorageKey, T extends TMlStorageMapped<K>>(
   key: K,
-  initValue?: TMlStorageMapped<K>
-): [TMlStorageMapped<K> | undefined, (value: TMlStorageMapped<K>) => void] {
-  const { value, setValue } = useContext(MlStorageContext);
+  initValue?: T
+): [
+  typeof initValue extends undefined
+    ? TMlStorageMapped<K> | undefined
+    : Exclude<TMlStorageMapped<K>, undefined>,
+  (value: TMlStorageMapped<K>) => void
+] {
+  const { value, setValue, removeValue } = useContext(MlStorageContext);
 
   const resultValue = useMemo(() => {
-    return (value?.[key] ?? initValue) as TMlStorageMapped<K>;
+    return (value?.[key] ?? initValue) as typeof initValue extends undefined
+      ? TMlStorageMapped<K> | undefined
+      : Exclude<TMlStorageMapped<K>, undefined>;
   }, [value, key, initValue]);
 
   const setVal = useCallback(
     (v: TMlStorageMapped<K>) => {
-      setValue(key, v);
+      if (isDefined(v)) {
```

Comment:
Do we want to be using `isDefined` here? a consumer of this function may want to explicitly set `null` as the stored value, for whatever reason.
I think we should only be checking for `undefined`

IMO `isDefined` is badly named as it considers `null` to be not defined.

Reply:
`localStorage` _**always**_  returns `null` for unknown keys or empty values, hence I assume it doesn't matter in this case

Reply:
Good point, ok